### PR TITLE
CEDS-2369 - implement cache clearance on start journey

### PR DESCRIPTION
**Pull request by SteveSugden**
I've done some experimenting with implementing an Action to clear the cache.  I needed to add a new endpoint to cause this to be triggered so I created one (POST /start) which just ensures the user is logged in, clears the cache and then redirects to the first step (Contact Address).

Only tested manually so far with two users in two different browsers.

### CODE REVIEW

File: app/connectors/Cache.scala [hunk 1]
```diff
@@ -35,6 +35,9 @@ class MongoCacheConnector @Inject()(val repository: CacheMapRepository) extends
 
   def getEntry[A](cacheId: String, key: String)(implicit hc: HeaderCarrier, fmt: Format[A]): Future[Option[A]] =
     fetch(cacheId).map(_.flatMap(_.getEntry(key)))
+
+  def remove()(implicit hc: HeaderCarrier): Unit =
+    repository.remove()
```
Comment (PatrykRudnicki) [inline app/connectors/Cache.scala:40]: Remove method probably needs to be based on the `cacheId`.

Comment (PatrykRudnicki) [reply app/connectors/Cache.scala:40]: That's not true, http-caching-client retrieve the cache ID from header carrier

Comment (PatrykRudnicki) [reply app/connectors/Cache.scala:40]: `repository.remove()` return `Future[HttpResponse]` not unit.

```
def remove()(implicit hc: HeaderCarrier, executionContext: ExecutionContext): Future[HttpResponse] =
    for {
      c      <- cacheId
      result <- delete(buildUri(defaultSource, c))
    } yield result
```

Can you change the return type?

File: app/connectors/Cache.scala [hunk 2]
```diff
@@ -44,4 +47,6 @@ trait Cache {
   def fetch(cacheId: String)(implicit hc: HeaderCarrier): Future[Option[CacheMap]]
 
   def getEntry[A](cacheId: String, key: String)(implicit hc: HeaderCarrier, fmt: Format[A]): Future[Option[A]]
+
+  def remove()(implicit hc: HeaderCarrier)
```
Comment (PatrykRudnicki) [inline app/connectors/Cache.scala:51]: This you'll need to have the return type which will be `Future[HttpResponse]`. For now it's treated like Unit which is not the same with coparison with `httpCacheClient`

File: app/controllers/actions/CacheDeleteAction.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class CacheDeleteActionImpl @Inject()(val dataCacheConnector: Cache, val exc: ExecutionContext) extends CacheDeleteAction {
+  override def filter[A](request: AuthenticatedRequest[A]): Future[Option[Result]] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+    dataCacheConnector.remove()
+    Future.successful(None)
```
Comment (PatrykRudnicki) [inline app/controllers/actions/CacheDeleteAction.scala:34]: After changes in cache you don't need to add Future.successful(None) you will need to just map to None if you want this return type

File: app/views/components/submit_button.scala.html [hunk 4]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  *@
 
-@()(implicit messages: Messages)
+@(labelKey: String = "common.continue")(implicit messages: Messages)
 
-  <button id="submit" class="button">@messages("common.continue")</button>
+  <button id="submit" class="button">@messages(labelKey)</button>
```
Comment (PatrykRudnicki) [inline app/views/components/submit_button.scala.html:19]: 👍

File: app/controllers/actions/CacheDeleteAction.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import javax.inject.{Inject, Singleton}
+import models.requests.AuthenticatedRequest
+import play.api.mvc.{ActionFilter, Result}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class CacheDeleteActionImpl @Inject()(val dataCacheConnector: Cache, val exc: ExecutionContext) extends CacheDeleteAction {
```
Comment (PatrykRudnicki) [inline app/controllers/actions/CacheDeleteAction.scala:29]: In general execution context can be here as implicit. In that case you don't need to pass it manually during invoking remove method

Comment (SteveSugden) [reply app/controllers/actions/CacheDeleteAction.scala]: Was worried about interfering with the required `override protected def executionContext: ExecutionContext` method.   But your suggestion is a good one so I've done it.

File: app/controllers/actions/CacheDeleteAction.scala [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import javax.inject.{Inject, Singleton}
+import models.requests.AuthenticatedRequest
+import play.api.mvc.{ActionFilter, Result}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class CacheDeleteActionImpl @Inject()(val dataCacheConnector: Cache, implicit val exc: ExecutionContext) extends CacheDeleteAction {
```
Comment (PatrykRudnicki) [inline app/controllers/actions/CacheDeleteAction.scala:29]: implicit we're putting in a separate brackets
`class CacheDeleteActionImpl @Inject()(val dataCacheConnector: Cache)(implicit val exc: ExecutionContext)`

but it's okay

Comment (SteveSugden) [reply app/controllers/actions/CacheDeleteAction.scala:29]: No, you are right.  I'm too quick to accept anything that compiles

**Pull request merged**